R/analysis.R

```diff
@@ -88,18 +90,67 @@ read_recording <- function(fileName) {
       input_line_number = seq_len(nrow(.)),
       time = end - start,
       concurrency = 1,
-      label = paste0(input_line_number, ":", event)
+      label = recording_item_labels(json)
     ) %>%
+    arrange(input_line_number) %>%
+    mutate(label = factor(label, levels = label, ordered = TRUE)) %>%
     select(
-      session_id, user_id, iteration,
+      # session_id, user_id, iteration,
       input_line_number,
       event,
       start, end, time,
-      concurrency,
-      label
+      # concurrency,
+      label,
+      json
     )
 }
 
+recording_item_labels <- function(x_list) {
+  shorten_url <- function(u) {
+    if (grepl("^/[^/]+/[^/]+/", u)) {
+      parts <- strsplit(u, "/")[[1]]
+      paste0(".../", parts[length(parts)])
+    } else {
+      u
+    }
+  }
+  ret <- c()
+  input_changes <- c()
+
+  prepend_line <- function(a, i) {
+    paste0(i, ") ", a)
+  }
+
+  for (i in seq_along(x_list)) {
+    x <- x_list[[i]]
+
+    new_label <- switch(x$type,
+      "REQ_HOME" = "Get: Homepage",
+      "REQ_GET" = paste0("Get: '", shorten_url(x$url), "'"),
+      "WS_OPEN" = "<Open Websocket>",
+      "WS_RECV_INIT" = "<Initialize Websocket>",
+      "WS_SEND" = {
+        if (i > 1 && identical(x_list[[i - 1]]$type, "WS_RECV_INIT")) {
+          "<Initialize Inputs>"
+        } else {
+          message <- jsonlite::fromJSON(x$message)
+          name_vals <- names(message$data)
+          visible_name_vals <- name_vals[!grepl("^\\.", name_vals)]
+          paste0("Set: ", paste0(visible_name_vals, collapse = ", "))
+        }
+      },
+      "WS_RECV" = {
+        message <- jsonlite::fromJSON(x$message)
+        paste0("Updated: ", paste0(names(message$values), collapse = ", "))
+      },
+      "WS_CLOSE" = "<Close Websocket>",
```

Comment:
These events don't seem to have a time associated with them in the waterfall* plot, do we even need to display them? If we do display them, I think it should just be "Session end" or "End" to avoid visual clutter.

*: I know that's not the correct name, is there a name for these?